### PR TITLE
[WIP][Core]The first PR for running worker in container

### PR DESCRIPTION
**Pull request by chenk008**
## Why are these changes needed?

We want to use container to resource limit, and use container image to support different dependencies.
This is the first PR to implements that feature. By using Podman, Worker process will started in container. 

Now runtime_env is serialized in cpp code, we can't reuse runtime_env to specify the contain image.We add container image option fields into `job_config`, and supports to specify the contain image. 

We add worker_resource into `WorkerCacheKey`, because worker_resource is one of worker’ characteristic, different actors have different worker_resource should not reuse worker process especially in container mode.

Start worker in container:

1. start ray with config 
```
ray start --head --port=6379 --system-config='{"worker_process_in_container_enabled":true}'
```

2 start job with container image
```
job_config = ray.job_config.JobConfig(worker_container_image="ray:latest")

ray.init(address="auto",job_config=job_config)

```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by caitengwei** [comment]
可以加一个 e2e 启动的case么？

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -420,3 +420,7 @@ RAY_CONFIG(int, publish_batch_size, 5000)
 /// The time where the subscriber connection is timed out in milliseconds.
 /// This is for the pubsub module.
 RAY_CONFIG(uint64_t, subscriber_timeout_ms, 30000)
+
+/// Feature flag to start worker process in individual container.
+/// Currently it only works on linux.
+RAY_CONFIG(bool, worker_process_in_container, false)
```
Comment (jovany-wang) [inline src/ray/common/ray_config_def.h:426]:
```suggestion
RAY_CONFIG(bool, worker_process_in_container_enabled, false)
```

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -420,3 +420,7 @@ RAY_CONFIG(int, publish_batch_size, 5000)
 /// The time where the subscriber connection is timed out in milliseconds.
 /// This is for the pubsub module.
 RAY_CONFIG(uint64_t, subscriber_timeout_ms, 30000)
+
+/// Feature flag to start worker process in individual container.
+/// Currently it only works on linux.
```
Comment (jovany-wang) [inline src/ray/common/ray_config_def.h:425]: It's better to check if this flag is only enabled on linux if it's only used on linux.

File: python/ray/job_config.py [hunk 3]
```diff
@@ -27,7 +27,8 @@ def __init__(self,
                  runtime_env=None,
                  client_job=False,
                  metadata=None,
-                 ray_namespace=None):
+                 ray_namespace=None,
+                 container_image=None):
```
Comment (jovany-wang) [inline python/ray/job_config.py:31]: container_image -> worker_container_image

File: src/ray/protobuf/gcs.proto [hunk 4]
```diff
@@ -289,6 +289,8 @@ message JobConfig {
   string serialized_runtime_env = 7;
   // An opaque kv store for job related metadata.
   map<string, string> metadata = 8;
+  // The worker container image.
+  string container_image = 9;
```
Comment (jovany-wang) [inline src/ray/protobuf/gcs.proto:293]: worker_container_image

File: src/ray/raylet/worker_pool.cc [hunk 5]
```diff
@@ -76,7 +78,10 @@ WorkerPool::WorkerPool(instrumented_io_context &io_service, const NodeID node_id
           num_initial_python_workers_for_first_job, maximum_startup_concurrency)),
       num_initial_python_workers_for_first_job_(num_initial_python_workers_for_first_job),
       periodical_runner_(io_service),
-      get_time_(get_time) {
+      get_time_(get_time),
+      worker_process_in_container_(worker_process_in_container),
+      temp_dir_(temp_dir),
+      session_dir_(session_dir) {
```
Comment (jovany-wang) [inline src/ray/raylet/worker_pool.cc:84]:
```suggestion
      session_dir_(std::move(session_dir)) {
```
or pass it as a reference.

File: src/ray/raylet/worker_pool.cc [hunk 6]
```diff
@@ -311,7 +325,13 @@ Process WorkerPool::StartWorkerProcess(
 
   // Start a process and measure the startup time.
   auto start = std::chrono::high_resolution_clock::now();
-  Process proc = StartProcess(worker_command_args, env);
+  Process proc;
+  std::string container_image = job_config->container_image();
+  if (worker_process_in_container_ && container_image != "") {
```
Comment (jovany-wang) [inline src/ray/raylet/worker_pool.cc:330]: How about `RAY_CHECK(!container_image.empty())` if it's enabled.

**Comment by jovany-wang** [comment]
I have a few of questions here:
1) How do we use this in production? Do we need to prepare image in every node?
2) How do validate that the image works with a correct version?
3) How does a user use the image? Does he need to build an image top on Ray-Worker-Image, and add his item to the images?

**Pull request closed without merging**